src/bytes.rs

```diff
@@ -1104,14 +1104,25 @@ unsafe fn release_shared(ptr: *mut Shared) {
     Box::from_raw(ptr);
 }
 
+#[cfg(miri)]
 fn ptr_map<F>(ptr: *mut u8, f: F) -> *mut u8
 where
     F: FnOnce(usize) -> usize,
 {
     let old_addr = ptr as usize;
     let new_addr = f(old_addr);
-    // this optimizes better than `ptr.wrapping_add(new_addr.wrapping_sub(old_addr))`
-    ptr.wrapping_sub(old_addr).wrapping_add(new_addr)
+    let diff = new_addr.wrapping_sub(old_addr);
+    ptr.wrapping_add(diff)
+}
+
+#[cfg(not(miri))]
+fn ptr_map<F>(ptr: *mut u8, f: F) -> *mut u8
+where
+    F: FnOnce(usize) -> usize,
+{
+    let old_addr = ptr as usize;
+    let new_addr = f(old_addr);
+    new_addr as *mut u8
```

Comment:
Perhaps this is a bit pedantic, but this really feels like it ought to have a comment on it, IMO. To someone who doesn't know the specific reason for this, it seems quite suspect that we only do the "correct" thing under Miri, and the "real" code does the thing that's not strict provenance compliant. A comment saying that the thing that's theoretically "correct" under strict provenance actually results in miscompiles in real life, and references the appropriate rustc issue etc.

Reply:
You aren't being pedantic. I added a link to this PR, which has links to the previous PR and the rust issue.